### DISCUSSION TITLE
Idempotent writes via HTTP ingress API

### DISCUSSION THREAD

**Discussion by zielmicha** [feldera/feldera#2278] [Q&A] [answered]
Is it possible to ingest data into Feldera via [HTTP API](https://www.feldera.com/docs/connectors/sources/http/) in a way that is reliable to network failures? E.g. if my HTTP request fails, how do I know whether it failed before the data was ingested (and I should retry) or after the data was ingested (and I shouldn't retry)?

**Comment by ryzhyk** [comment, marked as answer]
If HTTP is the only data source feeding the pipeline, you can achieve this by tracking the number of records ingested by the pipeline via the `/v0/pipelines/<pipeline_name>/stats` endpoint. The `global_metrics.total_input_records` counter tracks the total number of records ingested by the pipeline.  If the HTTP connector has successfully ingested the data, this counter will increment by the number of records in the HTTP request.

**Comment by lalithsuresh** [comment]
@zielmicha if your table has primary keys and you retry the HTTP request it should be idempotent (set semantics).
